core/src/main/java/org/infinispan/container/offheap/OffHeapEntryFactoryImpl.java

```diff
@@ -503,4 +523,105 @@ public long calculateSize(WrappedBytes key, WrappedBytes value, Metadata metadat
       }
       return UnpooledOffHeapMemoryAllocator.estimateSizeOverhead(totalSize + metadataSize);
    }
+
+   @Override
+   public long updateMaxIdle(long address, long currentTimeMillis) {
+      // 16 bytes for eviction if needed (optional)
+      // 8 bytes for linked pointer
+      long offset = evictionEnabled ? 24 : 8;
+
+      byte metadataType = MEMORY.getByte(address, offset);
+
+      if ((metadataType & (IMMORTAL + MORTAL)) != 0) {
+         return 0;
+      }
+
+      // skips over metadataType, hashCode
+      offset += 5;
+
+      int keySize = MEMORY.getInt(address, offset);
+      offset += 4;
+
+      boolean hasVersion = (metadataType & HAS_VERSION) != 0;
+
+      if ((metadataType & TRANSIENT) != 0) {
+         // Skip the metadataSize (if version present), valueSize and the keyBytes
+         offset += ((hasVersion ? 4 : 0) + 4 + keySize);
+         // Skip the max idle value
+         storeLongLittleEndian(address, offset + 8, currentTimeMillis);
+         return 0;
+      }
+      if ((metadataType & TRANSIENT_MORTAL) != 0) {
+         // Skip the metadataSize (if version present), valueSize and the keyBytes
+         offset += ((hasVersion ? 4 : 0) + 4 + keySize);
+         // Skip the lifespan/max idle values and created
+         storeLongLittleEndian(address, offset + 24, currentTimeMillis);
+         return 0;
+      }
+
+      byte[] metadataBytes = new byte[MEMORY.getInt(address, offset)];
+      int metadataSize = metadataBytes.length;
+      offset += 4;
+
+      int valueSize = MEMORY.getInt(address, offset);
+      offset += 4;
+
+      // skips over the actual key bytes
+      offset += keySize;
+
+      MEMORY.getBytes(address, offset, metadataBytes, 0, metadataSize);
+
+      Metadata metadata;
+      try {
+         metadata = (Metadata) marshaller.objectFromByteBuffer(metadataBytes);
+      } catch (IOException | ClassNotFoundException e) {
+         throw new CacheException(e);
+      }
+
+      Metadata newMetadata = metadata.builder()
+            .maxIdle(currentTimeMillis, TimeUnit.MILLISECONDS)
+            .build();
+
+      byte[] newMetadataBytes;
+      try {
+         newMetadataBytes = marshaller.objectToByteBuffer(newMetadata, metadataSize);
+      } catch (IOException e) {
+         throw new CacheException(e);
+      } catch (InterruptedException e) {
+         Thread.currentThread().interrupt();
+         throw new CacheException(e);
+      }
+
+      int newMetdataSize = newMetadataBytes.length;
+      if (newMetdataSize != metadataSize) {
```

Comment:
With metadata, we don't know for sure that the serialized form is the same with different timestamps. For example if they were doing some sort of UnsignedNumeric where it can vary in size.

Reply:
None of the `InternalCacheEntry.touch()` implementations change the metadata.

Reply:
That isn't the point. If you notice this branch is specifically when a CUSTOM or VERSION metadata is used. In either case we set the access time on the metadata. If the serialized form is different, we have to put it somewhere...

Reply:
How are you actually setting the access time on the metadata?

The last access timestamp only exists in `InternalMetadata`, which is an internal type, and since 10.0 it's never stored in the container/stores, it's only used with `cache.put()` to preserve the timestamps.

Reply:
The problem is calls like this https://github.com/infinispan/infinispan/blob/0309bf41bc84451836853a95f4fda22f02704c1b/core/src/main/java/org/infinispan/container/offheap/OffHeapEntryFactoryImpl.java#L581 

I don't quite understand the obsession with `InternalMetadata` as a user can supply any `Metadata` implementation. For example the `MetadataAPITest` tests that the custom metadata is stored and that it can be read back.

Reply:
That code is wrong: `maxIdle` is "maximum idle time of the entry in number of milliseconds", not the last access timestamp. `Metadata` just doesn't have any timestamps in it any more.

Reply:
That is the sole usage of last access timestamp is for maxIdle though.. Unless there is some other usage you can point me to, would be great to know.

Reply:
And also to point out the same type of thing is used here for the Metadata Builder, so we should make sure that is correct as well.

https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/cache/impl/CacheImpl.java#L1965

Reply:
Sorry I misunderstood your point earlier. That means that expiration with a custom metadata, which includes lifespan, has been broken forever in OFF_HEAP. This requires quite a bit of refactoring and changing the ordering of code to handle this assumption. The code currently assumes that if a custom metadata is used it doesn't need to persist the expiration metadata, however it does.

Reply:
I created https://issues.redhat.com/browse/ISPN-11277 to handle the custom metadata and off heap.

Reply:
Ok